src/raytracer.rs

```diff
@@ -1,90 +1,76 @@
-        let w_lights = t_lights.clone();
-        let w_num_cpus = num_cpus;
-        let w_i = chunk_num;
-        crossbeam_utils::scoped::scope(|scope| {
-            scope.spawn(move || {
-                process_part(w_cam, w_obj, w_lights,
-                             w_num_cpus, w_i, chunk);
+              -> Vec<u8> {
+
+    let mut pool = Pool::new(num_cpus::get() as u32);
+    let surf = Surface::new(cam.width, cam.height);
+
+    pool.scoped(|scope| {
+        for chunk in surf.divide(32, 32) {
+            scope.execute(move || {
+                process_part(cam, objects, lights, chunk);
             });
-        });
-        chunk_num += 1;
-    }
-    pixels
+        }
+    });
+    surf.pixels
 }
 
-fn process_part(cam: Arc<Camera>, objects: Arc<Vec<Box<Object>>>, lights: Arc<Vec<Box<Light>>>,
-                num_cpus: usize, part_num: usize, pixels: &mut [f64]) {
+fn process_part(cam: &Camera, objects: &Vec<Box<Object>>, lights: &Vec<Box<Light>>,
+                chunk: Division) {
     let aspect: f64 = cam.width as f64 / cam.height as f64;
     let angle = cam.fov.to_radians().tan();
     let inv_width = 1. / cam.width as f64;
     let inv_height = 1. / cam.height as f64;
     let right = cam.up.cross(&cam.dir).normalize();
-    let part_height = cam.height / num_cpus as u32;
-    for y in 0..part_height {
-        for x in 0..cam.width {
+    for yi in 0..chunk.h {
+        for xi in 0..chunk.w {
+            let x = xi + chunk.x0;
+            let y = yi + chunk.y0;
+
             let xx = right * (2. * (x as f64 + 0.5) * inv_width - 1.)
                 * angle * aspect;
-            let yy = cam.up * (1. - 2. * (y as f64 + (part_height as f64 * part_num as f64 + 0.5)) * inv_height)
+            let yy = cam.up * (1. - 2. * (y as f64 + 0.5) * inv_height)
                 * angle;
+
             let ray = Ray {
                 origin: cam.eye,
                 dir: (cam.dir + xx + yy).normalize()
             };
-            let mut color = trace(&ray, objects.clone(), lights.clone());
-            if color == Vec3::new(0., 0.5, 1.) {
-                color = Vec3::new(0., part_num as f64 / num_cpus as f64, 0.7);
+
+            let mut color = trace(&ray, objects, lights);
+            if color == Vec3::new(0., 0., 0.) {
+                color = Vec3::new(chunk.x0 as f64 / cam.width as f64, chunk.y0 as f64/ cam.height as f64, 0.7);
             }
-            let index = ((y * cam.width + x) * 3) as usize;
-            pixels[index] = color.x;
-            pixels[index + 1] = color.y;
-            pixels[index + 2] = color.z;
+            chunk.set_color(x, y, color);
         }
     }
 }
 
-fn trace(ray: &Ray, weak_objects: Arc<Vec<Box<Object>>>, weak_lights: Arc<Vec<Box<Light>>>) -> Vec3<f64> {
-    let objects = weak_objects;
-    let lights = weak_lights;
-
+fn trace(ray: &Ray, objects: &Vec<Box<Object>>, lights: &Vec<Box<Light>>) -> Vec3<f64> {
     let mut tnear = INFINITY;
     for i in 0..objects.len() {
```

Comment:
an iterator like `for i in &objects` should work here

Reply:
Indeed.

Reply:
That way it won't consume the `objects` whilst iterating. I'm not sure if it's necessary, really, but just in case.

Reply:
Ah, thanks.